tools/codegen/api/python.py

```diff
@@ -0,0 +1,987 @@
+        schema_formals: List[str] = \
+            list(map(lambda a: a.argument_str(method=self.method),
+                     self.arguments(skip_outputs=skip_outputs)))
+        positional_argc = len(self.input_args)
+        if len(schema_formals) > positional_argc:
+            schema_formals.insert(positional_argc, '*')
+
+        return f'{self.name}({", ".join(schema_formals)})'
+
+# The deprecated python signature involves some special logic, so create a
+# dedicated data model to store these extra properties.
+@dataclass(frozen=True)
+class PythonSignatureDeprecated(PythonSignature):
+    # We need keep the order of arguments in deprecated signature.
+    # Particularly, method signature might have 'self' not at the beginning, e.g.:
+    #   addmm(Scalar beta, Tensor self, Tensor mat1, Tensor mat2)
+    # When generating lambda function signature we need follow the exact order (even for method=True):
+    #   [](Scalar beta, const Tensor & self, const Tensor & mat1, const Tensor & mat2) -> Tensor
+    deprecated_args_names: Tuple[str]
+
+    # The deprecated signature might miss some arguments that the corresponding
+    # C++ signature expects. We need store the constant default values to pass in.
+    # For example:
+    #   [deprecate signature]: addmm(Scalar beta, Tensor self, Tensor mat1, Tensor mat2)
+    #   [func schema]: aten::addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
+    #   [func call]: self.addmm(mat1, mat2, beta, 1)
+    # We store ['self', 'mat1', 'mat2', 'beta', '1'] in this case.
+    deprecated_args_exprs: Tuple[str]
+
+    @property
+    def deprecated(self) -> bool:
+        return True
+
+    def signature_str(self, *, skip_outputs: bool = False) -> str:
+        return PythonSignature.signature_str(self, skip_outputs=skip_outputs) + '|deprecated'
+
+# C++ function dispatch is wrapped in a lambda function. The lambda function
+# has almost the same signature as the C++ function, only with some small
+# variants - see details below.
+# This data model is used to represent arguments of the lambda function
+# signature.
+@dataclass(frozen=True)
+class DispatchLambdaArgument:
+    name: str
+    type_str: str
+    cpp_type_str: str
+    is_out_arg: bool
+
+# To pass PyObjects arguments to C++ function (via the lambda wrapper),
+# we need first convert PyObjects into simple C++ objects. This work
+# is done by PythonArgParser.
+# This data model is used to represent the output of PythonArgParser.
+# It has 1-1 mapping with PythonArgument in PythonSignature.
+@dataclass(frozen=True)
+class PythonArgParserOutputExpr:
+    # argument name
+    name: str
+
+    # RHS expression to reference PythonArgParser output.
+    expr: str
+
+    # In some special cases we need create different expr, e.g.:
+    # '_r.isNone(1)' instead of '_r.tensor(1)'.
+    # TODO: maybe directly create 'isNone' form expr?
+    index: int
```

Comment:
Not exactly sure how `index: int` here relates to the comment above. How does knowing the index help you create a different expr?

Reply:
> Not exactly sure how `index: int` here relates to the comment above. How does knowing the index help you create a different expr?

Because these exprs all have index number in it, e.g.: '_r.isNone(1)' v.s. '_r.tensor(1)' they all have '1' in it which is the index. There are a few places calling 'isNone(i)' method.